x/gamm/pool-models/balancer/pool.go

```diff
@@ -714,20 +727,22 @@ func (p *Pool) CalcJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee s
 	// * numShares is how many shares are perfectly matched.
 	// * remainingTokensIn is how many coins we have left to join, that have not already been used.
 	// if remaining coins is empty, logic is done (we joined all tokensIn)
-	numShares, remainingTokensIn, err := cfmm_common.MaximalExactRatioJoin(p, sdk.Context{}, tokensIn)
+	numShares, tokensJoined, remainingTokensIn, err := p.CalcJoinPoolNoSwapShares(ctx, tokensIn, swapFee)
 	if err != nil {
 		return sdk.ZeroInt(), sdk.NewCoins(), err
 	}
-	if remainingTokensIn.Empty() {
-		tokensJoined = tokensIn
+
+	// safely ends the calculation if all input tokens are successfully LP'd
+	if tokensJoined.IsAnyGT(tokensIn) {
+		return sdk.ZeroInt(), sdk.NewCoins(), errors.New("an error has occurred, more coins joined than tokens passed in")
```

Comment:
Context: this check was included for single-asset joins at the end but for some reason not for all-asset joins (even though the method can terminate here). Added for consistency.

Reply:
This check already occurs inside `CalcJoinPoolNoSwapShares`. I think we might not need it here again

Reply:
However, if you follow my other suggestion of removing a 3-value return from `CalcJoinPoolNoSwapShares`, I think it makes sense to have it here instead